### PR TITLE
Server: Support custom TCP listener

### PR DESCRIPTION
**Pull request by aknuds1**
**What this PR does / why we need it**:
Refactor the Grafana server, so it can have a custom TCP listener. This is required by my [PR](https://github.com/grafana/grafana/pull/26517) for adding HTTP API tests, since it depends on starting Grafana server instances with random ports.

Also move the server itself to a package separate from `main`, so it can be instantiated from integration tests.

@bergquist has, as I understood him, OK-ed the integration testing approach in #26517, so this building block of said PR should be OK.

### CODE REVIEW

File: pkg/server/server.go [hunk 1]
```diff
@@ -100,13 +127,31 @@ func (s *Server) Run() (err error) {
 			continue
 		}
 
-		s.log.Debug("Initializing " + service.Name)
-
+		if cfg != nil {
+			if httpS, ok := service.Instance.(*api.HTTPServer); ok {
```
Comment (bergquist) [inline pkg/server/server.go:131]: feels a little bit hacky but I guess its ok. Perhaps you can add some comments here explaining this?

**Review by bergquist — APPROVED** [review]
Two comments that might be good to fix before merging. But neither is required. Up to you.

:+1:

**Comment by aknuds1** [comment]
Thanks @bergquist - will have a look at your comments!

File: pkg/server/server.go [hunk 2]
```diff
@@ -70,28 +89,34 @@ type Server struct {
...
-	HTTPServer    *api.HTTPServer       `inject:""`
+	HTTPServer *api.HTTPServer `inject:""`
 }
 
-// Run initializes and starts services. This will block until all services have
-// exited. To initiate shutdown, call the Shutdown method in another goroutine.
-func (s *Server) Run() (err error) {
+// init initializes the server and its services.
+func (s *Server) init(cfg *Config) error {
+	if s.isInitialized {
```
[the remark this answers is not in the document]
Comment (sakjur) [reply pkg/server/server.go:108]: Maybe a `sync.Once`?

> [*Once.]Do is intended for initialization that must be run exactly once.

https://golang.org/pkg/sync/#Once

[the remark this answers is not in the document]
Comment (aknuds1) [reply pkg/server/server.go:108]: Nice tip @sakjur.

[the remark this answers is not in the document]
Comment (aknuds1) [reply pkg/server/server.go:108]: The problem with sync.Once as I see it is that it doesn't allow for returning an error, we can also easily replicate what it does (checking an atomic int for performance before taking a lock).

[the remark this answers is not in the document]
Comment (aknuds1) [reply pkg/server/server.go:113]: Now protecting this behind mutex.

File: pkg/server/server.go (hunk 1 above)
Comment (aknuds1) [reply pkg/server/server.go:136]: Added comment.

**Pull request merged**